### PR TITLE
[Fetch Migration] Adding orchestration script for Fetch Migration steps

### PR DESCRIPTION
**Pull request by kartg**
### Description
This change adds an "orchestration" Python script that runs the Fetch Migration steps in order
1. The pre-migration script is run first to compare the indices on source and target clusters. A human-readable report is printed, and an updated Data Prepper pipeline YAML file is written to the Data Prepper `pipelines` directory
2.  Next, Data Prepper is kicked off as a [sub-process](https://docs.python.org/3/library/subprocess.html) to migrate data based on the output pipeline YAML file
3. After the Data Prepper process is kicked off, the monitoring script is run to poll DP's Prometheus metrics endpoint and determine when migration is complete and the process can be shut down

The orchestrator script takes multiple command-line inputs to enable execution of these steps. The Dockerfile definition has also been updated to use the orchestrator script as the entrypoint.
* Category: New feature

### Issues Resolved
N/A

### Testing
```
$ python -m coverage report --omit "*/tests/*"
Name                           Stmts   Miss  Cover
--------------------------------------------------
endpoint_info.py                   6      0   100%
fetch_orchestrator.py             24      0   100%
index_operations.py               35      0   100%
metadata_migration.py            123      1    99%
metadata_migration_params.py       7      0   100%
metadata_migration_result.py       5      0   100%
migration_monitor.py              59      0   100%
migration_monitor_params.py        5      0   100%
utils.py                          13      0   100%
--------------------------------------------------
TOTAL                            277      1    99%

```

### Check List
- [x] New functionality includes testing
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented
- [x] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: FetchMigration/Dockerfile
Comment (gregschohn) [inline FetchMigration/Dockerfile:18]: Since this will be situated within something else (which may also be situated within something else, because you never know) - you might want to disambiguate "orchestrator" with "fetch" (e.g. fetchOrchestrator.py)

What does ICT stand for?

File: FetchMigration/python/migration_monitor.py
Comment (gregschohn) [inline FetchMigration/python/migration_monitor.py:54]: consider prefacing these values w/ labels.
Second question - what does it mean to be inFlight and what does "no_part" mean?  From looking through the existing code, my bet is that it's sortof like a reverse countdown - something like how many of the original partitions no longer need to be moved?  Consider a clearer name to self document this field.

File: FetchMigration/python/migration_monitor.py
Comment (gregschohn) [inline FetchMigration/python/migration_monitor.py:71]: Consider renaming this to be more descriptive for the user.  e.g. poll_interval_seconds

File: FetchMigration/python/migration_monitor.py
Comment (gregschohn) [inline FetchMigration/python/migration_monitor.py:78]: Is there a reason NOT to log when there are no successful docs migrated?  It feels like there's a hole in the logging story w/out that too, especially if things get really messed up.

File: FetchMigration/python/migration_monitor.py
Comment (gregschohn) [inline FetchMigration/python/migration_monitor.py:89]: Would be nice to state whether it was successful or not here.

File: FetchMigration/python/migration_monitor_params.py
Comment (gregschohn) [inline FetchMigration/python/migration_monitor_params.py:7]: consider data_prepper_endpoint.  Readers might assume that you meant "data plane".

File: FetchMigration/python/orchestrator.py
Comment (gregschohn) [inline FetchMigration/python/orchestrator.py:18]: what will this output contain?  Will it be a report, logs?  It's fair for this component to leave that to DataPrepper if it's just stdout or labeled as output, but I'm still curious (so that as a reviewer, I can think about how the output needs to be shared/secured).

File: FetchMigration/python/orchestrator.py
Comment (gregschohn) [inline FetchMigration/python/orchestrator.py:21]: Please come up with a name that is more descriptive.  Everything is a migration, including the "pre" steps in many cases.  Say something like "index_config_migration" or "detecting_configurations", etc.

File: FetchMigration/python/orchestrator.py
Comment (gregschohn) [inline FetchMigration/python/orchestrator.py:24]: Do you want to print out the args that you're sending to it too?

File: FetchMigration/python/orchestrator.py
Comment (gregschohn) [inline FetchMigration/python/orchestrator.py:26]: how do you know that it was successful?  What happens when the code isn't successful?  I'm presuming that it throws, but I don't see a catch block that would explain to the user what has happened or what course of action to take next.

File: FetchMigration/python/orchestrator.py
Comment (gregschohn) [inline FetchMigration/python/orchestrator.py:33]: What signal does this send?  I'm hoping that its TERM.  If it doesn't respond to TERM w/in N seconds (maybe 10 or 30), it seems like sending a KILL (-9) is appropriate - though you might want to check that it didn't find more data in the interim and has begun to migrate it, if that's a possibility.  
Generally, this is going to be a difficult concurrency problem to solve and should be solved mostly in DataPrepper (you send a signal, it cleans up and handles its own health, data and consistency)

**Review by gregschohn — COMMENTED** [review]
Lots of naming suggestions & I have some concerns about corner cases around the process management of DataPrepper, some of which are probably outside the bounds of this PR.

File: FetchMigration/Dockerfile
Comment (kartg) [reply FetchMigration/Dockerfile:18]: Renamed `orchestrator.py` to `fetch_orchestrator.py`

ICT = Index Configuration Tool. This is left over from when the tool would only perform index configuration rather than the full fetch migration workflow. Changed references to `ICT` to `FM` instead.

File: FetchMigration/python/migration_monitor.py
Comment (kartg) [reply FetchMigration/python/migration_monitor.py:54]: * `in_flight` is a [counter published by Data Prepper](https://github.com/opensearch-project/data-prepper/blob/main/docs/monitoring.md#plugin-metrics) to describe the number of records in memory that have been pulled from the source plugin and are currently being processed by the sink plugin
* The `no_part` variables track the number of times the Data Prepper source coordination implementation attempted to fetch a  partition to move data for and found none pending. This is our indicator that data migration has finished and the DP pipeline is idle. The DP maintainers advised us to the track the _rate of change_ of this value rather than just a non-zero value, hence the `prev` counter.

Incorporated PR feedback by adding labels and renaming `no_part` variables to `no_partition` for clarity.

File: FetchMigration/python/migration_monitor.py
Comment (kartg) [reply FetchMigration/python/migration_monitor.py:89]: Reworded to `Migration monitor observed successful migration and idle pipeline, shutting down...`

File: FetchMigration/python/orchestrator.py
Comment (kartg) [reply FetchMigration/python/orchestrator.py:18]: `output_file` is the path to which `pre_migration.py` will write a modified Data Prepper pipeline YAML file after comparing indices between the source and target clusters. This pipeline YAML file ensures that only indices that are safe to migrate are created on the target cluster (and eventually migrated). The Data Prepper executable uses this `output_file` as its pipeline configuration.

File: FetchMigration/python/orchestrator.py
Comment (kartg) [reply FetchMigration/python/orchestrator.py:26]: Added an item to the backlog to improve this - https://opensearch.atlassian.net/browse/MIGRATIONS-1331

File: FetchMigration/python/orchestrator.py
Comment (kartg) [reply FetchMigration/python/orchestrator.py:33]: > What signal does this send? I'm hoping that its TERM

`terminate` [sends SIGTERM](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.terminate)

> .... and should be solved mostly in DataPrepper (you send a signal, it cleans up and handles its own health, data and consistency)

Data Prepper handles this already.

> If it doesn't respond to TERM w/in N seconds (maybe 10 or 30), it seems like sending a KILL (-9) is appropriate - though you might want to check that it didn't find more data in the interim and has begun to migrate it, if that's a possibility.

Added an item to the backlog to track this improvement - https://opensearch.atlassian.net/browse/MIGRATIONS-1331
Finding more data is not a concern since the source plugin will use scroll / PIT queries against the source

File: FetchMigration/python/orchestrator.py
Comment (kartg) [reply FetchMigration/python/orchestrator.py:24]: The Data Prepper process consumes its pipeline configuration from a file, which could contain credential information. I don't think we should output the contents of this file to the logs.

File: FetchMigration/python/orchestrator.py
Comment (kartg) [reply FetchMigration/python/orchestrator.py]: Changed the notion of "pre-migration" to "metadata migration" throughout

**Review by gregschohn — APPROVED** [review]

**Pull request merged**
